rfc/rfc-88/rfc-88.md

```diff
@@ -0,0 +1,596 @@
+seamlessly integrated. The Spark-Hudi integration leverages Spark's core read/write functionalities a lot. While these 
+functionalities are well-tested and performant in Spark using cases, they cannot be reused in Hudi's integration with other engines.
+
+As an example, Spark-Hudi integration uses `ParquetFileFormat` from Spark to build reader for parquet files, and uses 
+`ParquetWriteSupport` to build writer to handle parquet file writing. These reading/write infras can not be reused directly
+in other engine integrations without abstraction, e.g.,the Flink-Hudi module has to implement its own basic file readers/writers.
+
+### Separated Expression optimization
+Currently, the pushdown optimization for different engines do not share much in the integration between Expression and 
+MDT indices. For instance, there is `HoodieFileIndex` for spark scan to support partition pruning and data filtering, 
+which heavily relies on Spark DataFrame operations to evaluate Expressions against indices stats data(`SparkBaseIndexSupport`). 
+Meanwhile, for Flink engine, a separate Expression evaluator is also implemented to support evaluating pushed-down Flink 
+Expression against MDT stats data. There are quite a few common processing logics of index data between Flink FileIndex 
+and Spark FileIndex, which introduces difficulties both for the development and maintaining of functionalities about index optimization.
+
+## Goals
+To make the scope of the RFC more clear, we emphasize the goals here:
+* Propose a native schema as the authoritative schema for Hudi, making the type system more extensible and the semantic of data type clear and unified.
+* Abstract out the basic functionality for writers/readers based on data type abstraction, then each engine just implements the getter/setter logic, and eliminate the unnecessary Ser/De cost for the avro writer/read path.
+* Improve the expression abstractions and the integration with the Hudi file index and MDT indices, so that all these optimization can be shared between engines. Specifically, this should be used for expression indexes.
+
+## Design/Implementation
+
+### Abstraction of new Schema
+As elaborated above, to support comprehensive schema evolution, the basic data type and `InternalSchema` are introduced 
+and `InternalSchema#RecordType` can be losslessly converted to/from Avro schema.
+![schema_avro_schema_convert](schema_avro_schema_convert.png)
+Currently, `InternalSchema` is designed for internal use, and we are proposing a new Schema abstraction to serve as a 
+public API, while retaining the existing capabilities of `InternalSchema`.
+```java
+/**
+ * Definition of the schema for a Hudi table.
+ */
+@PublicAPIClass
+public class Schema implements Serializable {
+    private final RecordType record;
+    private int maxColumnId;
+    private long versionId;
+    private List<String> recordKeys;
```

Comment:
Can the keys be an expression of the columns? We seem to be assuming it's a simple key generator? (In hudi terminology)